src/ExactDiagonalization/solve.jl

```diff
@@ -0,0 +1,119 @@
+# a lazy type for iterating over coefficient vectors starting from a matrix
+struct LazyCoefficientVectors{VT,T} <: AbstractVector{VT}
+    m::Matrix{T}
+end
+function LazyCoefficientVectors(m::Matrix{T}) where {T}
+    return LazyCoefficientVectors{typeof(view(m, :, 1)),T}(m)
+end
+Base.size(lcv::LazyCoefficientVectors) = (size(lcv.m, 2), 1)
+Base.getindex(lcv::LazyCoefficientVectors, i::Int) = view(lcv.m, :, i)
+
+# a lazy type for iterating over coefficient vectors starting from a vector of DVecs
+struct LazyCoefficientVectorsDVecs{T,VDV<:Vector{<:AbstractDVec},B} <: AbstractVector{T}
+    vecs::VDV
+    basis::B
+end
+function LazyCoefficientVectorsDVecs(vecs, basis)
+    T = valtype(vecs[1])
+    return LazyCoefficientVectorsDVecs{T,typeof(vecs),typeof(basis)}(vecs, basis)
+end
+Base.size(v::LazyCoefficientVectorsDVecs) = (length(v.vecs),)
+function Base.getindex(lcv::LazyCoefficientVectorsDVecs{T}, i::Int) where {T}
+    return T[lcv.vecs[i][address] for address in lcv.basis]
+end
+
+# lazy type for iterating over constructed DVecs
+struct LazyDVecs{DV,LCV,B} <: AbstractVector{DV}
+    vs::LCV
+    basis::B
+end
+function LazyDVecs(vs::LCV, basis::B) where {LCV,B}
+    return LazyDVecs{typeof(DVec(zip(basis, vs[1]))),LCV,B}(vs, basis)
+end
+Base.size(lv::LazyDVecs) = size(lv.vs)
+Base.getindex(lv::LazyDVecs, i::Int) = DVec(zip(lv.basis, lv.vs[i]))
```

Comment:
I'd prefer `PDVec`s here as they're probably going to be used to compute operator dot products and such.
```suggestion
function LazyDVecs(vs::LCV, basis::B) where {LCV,B}
    return LazyDVecs{typeof(PDVec(zip(basis, vs[1]))),LCV,B}(vs, basis)
end
Base.size(lv::LazyDVecs) = size(lv.vs)
Base.getindex(lv::LazyDVecs, i::Int) = PDVec(zip(lv.basis, lv.vs[i]))
```